keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,563 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Future Work
+
+After OpenAPI v3 is implemented, we can start working on updating clients to consume OpenAPI v3. A separate KEP will be published for that.
+
+### Risks and Mitigations
+
+Aggregation for OpenAPI v2 already consumes a non-negligible amount of resources, naively adding in an aggregation for v3 will double the workload when the OpenAPI needs to be re-aggregated. Splitting by group partially mitigates this problem because when a resource is changed, only a small subset of groups will need to be reaggregated compared to the current behavior of re-aggregating everything.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Paths
+
+The overarching `/openapi/v3` endpoint will contain the list of paths (groups) available and serve as a discovery endpoint. Clients can then choose the group(s) to fetch and send additional requests.
+
+/openapi/v3
+
+```json
+{
+   "Paths" : [
+      "api",
+      "api/v1",
+      "apis",
+      "apis/admissionregistration.k8s.io",
+      "apis/apiextensions.k8s.io",
+      "apis/apps",
+      "apis/authentication.k8s.io",
+      ...
+   ]
+}
+```
+
+Based on the provided group, clients can then request `openapi/v3/apis/apps`, `/openapi/v3/api/networking.k8s.io` and etc. These leaf node specs are self contained OpenAPI v3 specs and include all referenced types.
+
+### Controllers
+
+#### OpenAPI Builder
+
+The OpenAPI builder for CRDs converts the v3 validation schema into a v2 schema. We will keep the conversion to the v2 schema for backwards compatibility, but also generate the v3 schema to publish at the new v3 endpoint.
+
+#### Proto Models & ETags
+
+Kubernetes publishes the openapi schema in both JSON and Protobuf. v3 schemas will include the same mechanisms as v2, publishing a protobuf model of the schema, as well as the corresponding etags for caching. In order to publish the schema based on groups, the CRDs will be grouped by group and published only at the endpoint for their specific group.
+
+#### Aggregator
+
+The aggregator has a mapping of all the APIServices and refreshes the aggregated spec if a single spec or CRD changes. Since the underlying api services will publish resources based on group, the aggregator must be able to discover all group endpoints for each APIService. This will increase the number of requests on bootstrap and subsequently the number of internal requests on a schema change. However, since aggregation is done by group, only a small subset of groups will be invalidated and reaggregated.
+
```

Comment:
This section misses a version skew discussion: old apiservers won't have v3 schemas served. Will kube-apiserver migrate them on-the-fly? Is that even possible?

Reply:
Added a small section for version skew. I feel like we can't do anything meaningful with the data from the old apiservers for v3. v2 will continue to be published and aggregated so that will be up to date. Given the v2 spec, we can't exactly retrieve the v3 spec without misrepresenting some information. I think we should drop the v3 publishing for old apiservers until they are upgraded.

Reply:
see my other comment. I tend to agree. Needs some explanation of the reasoning. Basically: why are the client we have in mind happy if v3 group-versions are missing?

Reply:
hmm, where is that dicussion?

Reply:
ic, further down under version skew.